File: src/BasicReadLHE.cpp
/** 
 *  @file   BasicReadLHE.cpp
 *  @brief  Contains simplest example on how to read data from .lhe file obtained from MadGraph with HepMC3
 *
 *  This file is a part of a project mg5-crossec-course (https://github.com/Sergeyir/mg5-crossec-course).
 *
 *  @author Sergei Antsupov (antsupov0124@gmail.com)
 **/
#ifndef BASIC_READ_LHE_CPP
#define BASIC_READ_LHE_CPP

#include "BasicReadLHE.hpp"

int main(int argc, char **argv)
{
   // printing info on usage and exiting the program if number of parameters is incorrect
   if (argc != 2) 
   {
      std::cout << "\033[1m\033[31mError:\033[0m Expected 1 parameter while " + 
                   std::to_string(argc - 1) + " parameter(s) were provided.\n"\
                   "Usage: bin/BasicReadLHE input_file_name output_file_name" << std::endl;
      return 1;
   }

   // Checking if specified file exist
   if (!std::filesystem::exists(argv[1]))
   {
      std::cout << "\033[1m\033[31mError:\033[0m File " << argv[1] << " does not exist" << std::endl;
      return 1;
   }

   std::cout << "Reading from file " << argv[1] << std::endl;

   // Initializing reader for LHE file. Reader is used to access information stored in .lhe file
   LHEF::Reader reader(argv[1]);

   unsigned long numberOfEvents = 0;
   while (reader.readEvent())
   {
      numberOfEvents++;

      if (numberOfEvents % 1000 == 0) std::cout << "Event " << numberOfEvents << "\r" << std::endl;

      // HepMC needs this object to store the event, tags, XML, and other data into
      std::shared_ptr<HepMC3::HEPEUPAttribute> hepe = std::make_shared<HepMC3::HEPEUPAttribute>();

      if (reader.outsideBlock.length())
      {
         // Each event may have non-standard information outside LHEF event 
         hepe->tags = LHEF::XMLTag::findXMLTags(reader.outsideBlock);
      }
      // Storing reader event data into HepMC3::HEPEUPAttribute object
      hepe->hepeup = reader.hepeup;

      if (numberOfEvents == 1)
      {
         // Iterating over every particles in the current event
         for (int i = 0; i < hepe->hepeup.NUP; i++)
         {
            HepMC3::GenParticlePtr particle = 
               std::make_shared<HepMC3::GenParticle>(hepe->momentum(i), 
                                                     hepe->hepeup.IDUP[i], 
                                                     hepe->hepeup.ISTUP[i]);

            // GenParticle and FourVector methods can be seen here
            // https://dayabay.bnl.gov/dox/HepMC/html/classHepMC_1_1GenParticle.html
            HepMC3::FourVector fourVec = particle->momentum();

            std::cout << particle->pid() << " " << fourVec.e() << " " << 
                         fourVec.px() << " " << fourVec.py() << " " << fourVec.pz() << std::endl;
         }
      }
   }
   
   // Find the measurement units for the total cross section
   std::cout << "Total cross section: " << GetTotCrossSection(argv[1]) << std::endl;

   return 0;
}

double GetTotCrossSection(const std::string& fileName)
{
   std::ifstream file(fileName);

   std::regex re(R"(^#\s*Integrated weight \(pb\)\s*:\s*([0-9.+-eE]+))");

   std::string line;
   while (std::getline(file, line))
   {
      std::smatch m;
      if (std::regex_search(line, m, re))
      {
         return std::stod(m[1].str());
      }
      // stopping if header was passed
      if(line.find("<event") != std::string::npos) break;
   }

   std::cout << "\033[1m\033[31mError:\033[0m cross-section comment "\
                "not found in file" + fileName << std::endl;
   exit(1);
}

#endif /* BASIC_READ_LHE_CPP */

File: scripts/basic_read_lhe.py
#!/usr/bin/env python3

#  @file   basic_read_lhe.py
#  @brief  Contains simplest example on how to read data from .lhe file obtained from MadGraph with HepMC3
#
#  This file is a part of a project genjets-course (https://github.com/Sergeyir/genjets-course).
#
#  @author Sergei Antsupov (antsupov0124@gmail.com)

import os
import re
import sys
import argparse

import ROOT
from pyHepMC3 import HepMC3, LHEF

"""
Returns the total cross section if the comment about it exists the specified .lhe file
Otherwise error will be printed and the program will be terminated
"""
def get_tot_cross_section(file_name: str) -> float:
    pattern = re.compile(r"^\s*#\s*Integrated weight \(pb\)\s*:\s*([0-9.+\-eE]+)")
    with open(file_name, 'r') as f:
        for line in f:
            m = pattern.search(line)
            if m:
                return float(m.group(1))
            if "<event" in line:
                break

    print(f"\033[1m\033[31mError:\033[0m cross-section comment not found in file {filename}", 
          file=sys.stderr)
    sys.exit(1)

def read_lhe(input_file_name : str) -> int :
    # checking if specified input file exists
    if not os.path.isfile(input_file_name):
        print(f"\033[1;31mError:\033[0m file {input_file_name} was not found")
        return 1

    # Initializing reader for .lhe file. Reader is used to access information in .lhe file
    reader = LHEF.Reader(input_file_name)

    number_of_events = 0

    while reader.readEvent() :
        number_of_events += 1

        if number_of_events % 1000 == 0 :
            print(f"Event {number_of_events}\r")

        # HepMC needs this object to store the event, tags, XML, and other data into
        hepe = HepMC3.HEPEUPAttribute()

        # Each event may have non-standard information outside LHEF event 
        if len(reader.outsideBlock) :
            hepe.tags = LHEF.XMLTag.findXMLTags(reader.outsideBlock)

        # Storing reader event data into HepMC3.HEPEUPAttribute object
        hepe.hepeup = reader.hepeup

        # Iteration over every particle in the current event
        for i in range(hepe.hepeup.NUP) :
            particle = HepMC3.GenParticle(hepe.momentum(i), 
                                          hepe.hepeup.IDUP[i], 
                                          hepe.hepeup.ISTUP[i])
            # GenParticle and FourVector methods can be seen here
            # https://dayabay.bnl.gov/dox/HepMC/html/classHepMC_1_1GenParticle.html
            four_vec = particle.momentum()

            print(particle.pid(), four_vec.e(), four_vec.px(), four_vec.py(), four_vec.pz())

    print("Total cross section:", get_tot_cross_section(input_file_name))
    return 0

if __name__ == "__main__":
    # parsing input arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", "--input", type=str)

    args = parser.parse_args()
    input_file_name = args.input

    if not input_file_name:
        print('\033[1;31mError:\033[0m .lhe input file must be specified; use option -i or --input')
        sys(exit(1))

    sys.exit(read_lhe(input_file_name))
